packages/react/src/experimental/OneDataCollection/useSelectable.ts

```diff
@@ -0,0 +1,133 @@
+    const unselected = new Map()
+    itemsState.forEach((value, id) => {
+      if (value.checked) {
+        selected.set(id, value.item)
+      } else {
+        unselected.set(id, value.item)
+      }
+    })
+    return [selected, unselected]
+  }, [itemsState])
+
+  const selectedCount: number = selectedItems.size
+  const unselectedCount: number = unselectedItems.size
+
+  const [allSelectedCheck, setAllSelectedCheck] = useState(false)
+
+  const handleItemStateChange = (item: Record, checked: boolean) => {
+    const id = source.selectable && source.selectable(item)
+    if (id === undefined) {
+      return
+    }
+
+    itemsState.set(id, { item, checked })
+
+    setItemsState(new Map(itemsState))
+  }
+
+  const handleSelectAll = (checked: boolean) => {
+    setAllSelectedCheck(checked)
+    if (checked) {
+      data.forEach((item) => {
+        handleItemStateChange(item, true)
+      })
+    } else {
+      // We need to reset the items state to the initial state
+      setItemsState(
+        new Map(
+          Array.from(itemsState.entries()).map(([id, value]) => {
+            return [id, { item: value.item, checked: false }]
+          })
+        )
+      )
+    }
+  }
+
+  // Try to determine if all items are selected when we select one by one
+  // If there is pagination, we need to check if the selected items are less than the total number of items
+  // If there is no pagination, we need to check if the selected items are less than the total number of items
+  const areAllKnownItemsSelected =
+    (paginationInfo && selectedCount === paginationInfo.total) ||
+    (!paginationInfo && selectedCount === data.length)
+
+  const isPartiallySelected = selectedCount > 0 && unselectedCount > 0
+
+  const isAllSelected = allSelectedCheck || areAllKnownItemsSelected
+
+  // If the filters change, we need to reset the selected items
+  useEffect(() => {
+    handleSelectAll(false)
+    setItemsState(new Map())
+  }, [source.currentFilters])
+
+  // If the data changes, we need to update new items that are added to the data
+  // If the data changes, we need to update new items that are added to the data
+  useEffect(() => {
+    if (isAllSelected) {
+      data.forEach((item) => {
+        const id = source.selectable && source.selectable(item)
+        // If the item was loaded before, we can't change the state
+        if (id !== undefined && !itemsState.has(id)) {
+          handleItemStateChange(item, true)
+        }
+      })
+    }
+  }, [data, isAllSelected])
+
+  // Control the allSelectedCheck state
+  // If all items are selected, we need to set the allSelectedCheck state to true
+  // If there are no selected items, we need to set the allSelectedCheck state to false
+  // If some items are selected, we need to keep the state as is to know if we will need to check the next page items
+  useEffect(() => {
+    if (areAllKnownItemsSelected) {
+      setAllSelectedCheck(true)
+    }
+    if (selectedCount === 0) {
```

Comment:
should we have it as `else if`? Could it be that we set true and then false in some circumstances?

Reply:
yes, agree, clearer code